src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java

```diff
@@ -200,12 +197,43 @@ public final void unregisterSession(final Session session) {
         }
 
         try {
+
+            final CountDownLatch latch = new CountDownLatch(zkSessions.size());
+            final Map<String, Session> resultSessions = new HashMap<>();
             for (final String sessionId : zkSessions) {
-                final int weight = Integer.parseInt(new String(getCurator().getData()
-                        .forPath(getSubscriptionPath("/sessions/" + sessionId)), UTF_8));
-                sessions.add(new Session(sessionId, weight));
+                getCurator().getData().inBackground((client, event) -> {
+                    try {
+                        if (event.getResultCode() == KeeperException.Code.OK.intValue()) {
+                            final Integer data = Integer.parseInt(new String(event.getData(), UTF_8));
+                            synchronized (resultSessions) {
+                                resultSessions.put(sessionId, new Session(sessionId, data));
+                            }
+                        } else {
+                            getLog().warn(
+                                    "Failed to get session {} information from zk. status code: {}",
+                                    sessionId,
+                                    event.getResultCode());
+                        }
+                    } finally {
+                        latch.countDown();
+                    }
+                }).forPath(getSubscriptionPath("/sessions/" + sessionId));
             }
-            return sessions.toArray(new Session[sessions.size()]);
+            if (latch.await(10, TimeUnit.SECONDS)) {
```

Comment:
I think 10 seconds is too much to wait, at least I expect one call to zk be less than 1 second, could you elaborate on that ?